### PR TITLE
Add .gitlab-ci.yml template

### PR DESCRIPTION
**Pull request by jeanphi-baconnais**
This is a first version of a GitLabCI template, using maven. I need to verify that this file allows to generate good images.

I create this PR in draft because i have an error on the unit test which find variables `maven.version` but I failed to resolve it.

### CODE REVIEW

**Review by gsmet — CHANGES_REQUESTED** [review]
So first, sorry for the delay. I think this is good work and it definitely has values. We should probably have the same thing for GitHub Actions.

I added a small comment there but nothing big.

Quick question: how is it triggered when creating a project?

**Comment by jeanphibaconnais** [comment]
Hi, no worry for the delay.
This pipeline runs after each commit. I didn't put specific rules because it can depend on each project owner. Do you think I can specify one by default ? For example basically run this pipeline only when we push on master ? 
I will check quickly your comment about native-deploy job.

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/gitlab-ci/base/.gitlab-ci.tpl.qute.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  needs: ["📦 native-build-maven"]
+  {#else}
+  needs: ["📦 native-build-gradle"]
+  {/if}
+  image:
+    name: gcr.io/kaniko-project/executor:debug
+    entrypoint: [""]
+  script:
+    - echo "{\"auths\":{\"$CI_REGISTRY\":{\"auth\":\"$(printf "%s:%s" "$CI_REGISTRY_USER" "$CI_REGISTRY_PASSWORD" | base64 | tr -d '\n')\" } } }" > /kaniko/.docker/config.json
+    - /kaniko/executor --context $CI_PROJECT_DIR --dockerfile $CI_PROJECT_DIR/src/main/docker/Dockerfile.native-distroless --destination $CI_REGISTRY_IMAGE:native
```
[the remark this answers is not in the document]
Comment (jeanphibaconnais) [reply independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/gitlab-ci/base/.gitlab-ci.tpl.qute.yml:116]: Ok I remember. The Dockerfile for native image contains user management. And the 'debug' version of Kaniko image seems to have a problem. I saw many discussions on this point.
So I add a native-deploy job but with classics docker commands. 

I added a workflow to run pipeline only on default branch. I validate my tests and I push it.

**Comment by maxandersen** [comment]
We have made similar for github actions but the main reason we didn't add it is we were missing ways to control when to add these files + it is necessary to setup tokens/secrets for things to work ( at least for deployment parts ).

So same question as @gsmet how is this enabled at project create time ? Is it just "always on" or how ?

/cc @ia3andy for codestarts aspect.

File: independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/gitlab-ci/base/.gitlab-ci.tpl.qute.yml [hunk 2]
```diff
@@ -0,0 +1,141 @@
+#
+# Template GitLab CI to build jvm, native and native-distroless images.
```
Comment (maxandersen) [inline independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/tooling/gitlab-ci/base/.gitlab-ci.tpl.qute.yml:2]: we really need to find a way to parameterize what approach is taken here as distroless isn't the only approach.

**Comment by ia3andy** [comment]
@jeanphibaconnais  How is meant to be activated by users? is it meant for a future support of GitLab on code.quarkus.io?

**Comment by maxandersen** [comment]
I'm wondering if we should have a separate notion of generation of these kind of files?

The content of gitlab.yml, .github/workflows, tekton etc. is highly dependent on what kind of build you are interested in having, what base images/project/product you are using. Thus it changes dependent on what your project is targeting.

Same or similar issue shows up with Dockerfile - it is statically generated and goes easily out of date.

Could we generate these based on the projects current target values at will rather than statically once at project creation ?

**Comment by maxandersen** [comment]
@jeanphibaconnais its about how the gitlab file will be used. It's how users tell it is necessary to have gitlab file generated. 

Afaics it will always be included which is not great given we got plenty of pushback when considering adding other files which would not be used by majority of users.

**Comment by maxandersen** [comment]
Sorry missed you last comment. 

Doing a “fake” gitlab extension is a way one could do today. Just feels “wrong” but happy to explore.

**Comment by jeanphibaconnais** [comment]
Hi @maxandersen , I review your comment and in this issue related https://github.com/quarkusio/quarkus/issues/24502 a discussion will be made. Coping the extension kubernetes is a good way to do ?

**Comment by Marco-Cocco** [comment]
Is there any update here? I'm in need of a proper approach for Gitlab CI\CD.
It's unreal that such a framework is still not under the belt of Quarkus, am I missing something?

**Comment by gsmet** [comment]
I'm closing this one as we haven't made progress since 2022.

If at some point, someone can contribute default GitLab CI workflows that could be distributed by default (I think that's the hard part), please open another PR.

Thanks!

**Comment by jeanphi-baconnais** [comment]
Hello @gsmet , when you said a default GitLab CI, that means a qute template as I started to work on? https://github.com/quarkusio/quarkus/pull/21980/files#diff-6d47b1f672b15e37404d79d725b882d10a4924d82ef4789eed48b2c2c05b2a6f 

If this is, I can try to initialize something from my first approach 🙂

**Comment by maxandersen** [comment]
this is related to #24502 

@gsmet not sure what you mean by distributing by default?

the main thing we've been missing how to choose gitlab vs github vs tekton vs ...which aren't necessary tied to an extension.

**Pull request closed without merging**